### PR TITLE
Optionality check in nyam2nxdl

### PR DESCRIPTION
**Pull request by RubelMozumder**

### CODE REVIEW

**Comment by RubelMozumder** [comment]
Implementation of changes:
   1. By default all NeXus attributes, fields and groups will have 'optional' attributes if no 'optional' attributes are rendered.
   2. By default 'optional' attributes is 'false' for the application and 'true' for the base definition.
   3. By default the dim element will get the 'required' attribute and value 'true' (app def) and 'false' (base def).
   4. TEST: The 'test_attributes()' test would benefit the above properties.

File: pynxtools/nyaml2nxdl/nyaml2nxdl_backward_tools.py [hunk 1]
```diff
@@ -538,7 +540,7 @@ def handle_dimension(self, depth, node, file_out):
           and attributes of dim has been handled inside this function here.
         """
         # pylint: disable=consider-using-f-string
-        possible_dim_attrs = ['ref', 'optional', 'recommended',
+        possible_dim_attrs = ['ref', 'recommended',
```
Comment (domna) [inline pynxtools/nyaml2nxdl/nyaml2nxdl_backward_tools.py:543]: recommended can be removed as well

File: pynxtools/nyaml2nxdl/nyaml2nxdl_forward_tools.py [hunk 2]
```diff
@@ -1106,6 +1154,7 @@ def nyaml2nxdl(input_file: str, out_file, verbose: bool):
         (lin_annot, line_loc) = post_comment.get_line_info()
         xml_handle_comment(xml_root, lin_annot, line_loc)
 
+    check_for_default_attribute_and_value(xml_root)
```
Comment (domna) [inline pynxtools/nyaml2nxdl/nyaml2nxdl_forward_tools.py:1157]: Defaults should not be written explicitly

**Review by domna — APPROVED** [review]

File: tests/data/nyaml2nxdl/Ref_NXellipsometry-docCheck.nxdl.xml [hunk 3]
```diff
@@ -66,7 +66,7 @@
          - https://www.angstromadvanced.com/resource.asp
          - https://pypolar.readthedocs.io/en/latest/
     </doc>
-    <group type="NXentry">
+    <group type="NXentry" optional="false">
```
Comment (sanbrock) [inline tests/data/nyaml2nxdl/Ref_NXellipsometry-docCheck.nxdl.xml:69]: should not the optionality be skipped in this nxdl.xml file?

File: tests/data/nyaml2nxdl/Ref_NXnested_symbols.nxdl.xml [hunk 4]
```diff
@@ -52,7 +52,7 @@
          Test case for verifying handling of symbols inside a nexus class in nested
          layers of the hierarchy
     </doc>
-    <group type="NXsample">
+    <group type="NXsample" optional="true">
```
Comment (sanbrock) [inline tests/data/nyaml2nxdl/Ref_NXnested_symbols.nxdl.xml:55]: same question here

File: tests/data/nyaml2nxdl/Ref_NXtest_links.nxdl.xml [hunk 5]
```diff
@@ -21,11 +21,11 @@
 #
 # For further information, see http://www.nexusformat.org
 -->
-<definition xmlns="http://definition.nexusformat.org/nxdl/3.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" category="base" type="group" extends="NXobject" name="NXentry" xsi:schemaLocation="http://definition.nexusformat.org/nxdl/3.1 ../nxdl.xsd">
+<definition xmlns="http://definition.nexusformat.org/nxdl/3.1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" category="base" name="NXentry" extends="NXobject" type="group" xsi:schemaLocation="http://definition.nexusformat.org/nxdl/3.1 ../nxdl.xsd">
     <doc>
          Test case for verifying that the parser can handle links correctly.
     </doc>
-    <group type="NXdata">
+    <group type="NXdata" optional="true">
```
Comment (sanbrock) [inline tests/data/nyaml2nxdl/Ref_NXtest_links.nxdl.xml:28]: and here

**Review by sanbrock — CHANGES_REQUESTED** [review]
Reference xml files still have unnecessary optional properties

File: tests/data/nyaml2nxdl/Ref_NXtest_links.nxdl.xml (hunk 5 above)
Comment (RubelMozumder) [reply tests/data/nyaml2nxdl/Ref_NXtest_links.nxdl.xml:28]: Removed!

File: tests/data/nyaml2nxdl/Ref_NXellipsometry-docCheck.nxdl.xml (hunk 3 above)
Comment (RubelMozumder) [reply tests/data/nyaml2nxdl/Ref_NXellipsometry-docCheck.nxdl.xml:69]: Removed!

File: pynxtools/nyaml2nxdl/nyaml2nxdl.py [hunk 6]
```diff
@@ -36,6 +36,9 @@
 
 DEPTH_SIZE = "    "
```
[the remark this answers is not in the document]
Comment (RubelMozumder) [reply pynxtools/nyaml2nxdl/nyaml2nxdl.py:37]: Editted!

**Comment by RubelMozumder** [comment]
> Reference xml files still have unnecessary optional properties

@sanbrock, all the auto-generated 'optional' has been removed. Still, some of the 'optional' keyword are left, they are not auto-generated but rather from changes of yaml.

**Review by sanbrock — APPROVED** [review]
LGTM

**Pull request merged**